plasmapy/simulation/particle_tracker.py

```diff
@@ -0,0 +1,842 @@
+        """
+        ...
+
+
+class TimeElapsedStopCondition(AbstractStopCondition):
+    """Stop condition corresponding to the elapsed time of a ParticleTracker."""
+
+    def __init__(self, stop_time: u.Quantity):
+        self._particle_tracker = None
+        self.stop_time = stop_time.to(u.s).value
+
+    @property
+    def require_uniform_dt(self):
+        """The elapsed time stop condition requires a uniform time step
+        to stop all particles at the same time.
+        """
+        return True
+
+    @property
+    def progress_description(self):
+        """The time elapsed stop condition depends on elapsed time,
+        therefore the relevant quantity is time remaining.
+        """
+        return "Time remaining"
+
+    @property
+    def units(self):
+        """The units for the time elapsed condition have the units of seconds."""
+
+        return "seconds"
+
+    @property
+    def is_finished(self):
+        """Conclude the simulation if all particles have been tracked over the specified stop time."""
+        return self._particle_tracker.time >= self.stop_time
+
+    @property
+    def progress(self):
+        """Return the current time step of the simulation."""
+        return self._particle_tracker.time
+
+    @property
+    def total(self):
+        """Return the total amount of time over which the particles are tracked."""
+        return self.stop_time
+
+
+class NoParticlesOnGridsStoppingCondition(AbstractStopCondition):
+    """Stopping condition corresponding to stopping the simulation when all particles have exited the grid."""
+
+    def __init__(self):
+        self._particle_tracker = None
+
+    @property
+    def require_uniform_dt(self):
+        """The no field stopping condition does not require a uniform time step."""
+        return False
+
+    @property
+    def progress_description(self):
+        """The progress meter is described in terms of the fraction of particles still on the grid."""
+        return "Number of particles still on grid"
+
+    @property
+    def units(self):
+        """The progress meter is described in terms of the fraction of particles still on the grid."""
+        return "Particles"
+
+    @property
+    def is_finished(self):
+        """The simulation is finished when no more particles are on any grids."""
+        is_not_on_grid = self._particle_tracker.on_grid == 0
+
+        return is_not_on_grid.all() and self._particle_tracker.iteration_number > 0
+
+    @property
+    def progress(self):
+        """The progress of the simulation is measured by how many particles are no longer on a grid."""
+        is_not_on_grid = self._particle_tracker.on_grid == 0
+
+        return is_not_on_grid.sum()
+
+    @property
+    def total(self):
+        """The progress of the simulation is measured against the total number of particles in the simulation."""
+        return self._particle_tracker.nparticles
+
+
+class AbstractSaveRoutine(ABC):
```

Comment:
Looking at the `adaptive_dt`, it occurs to me that we require a saving condition be provided, but what if the user doesn't care to save the data? We need a `NoSaveRoutine` that just does nothing when the hook is called. That would be a natural keyword default if the user doesn't specify a save routine.